pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarChannelInitializer.java

```diff
@@ -209,5 +209,33 @@ CompletableFuture<Channel> initSocks5IfConfig(Channel ch) {
 
         return initSocks5Future;
     }
+
+    CompletableFuture<Channel> initializeClientCnx(Channel ch,
+                                                   InetSocketAddress logicalAddress,
+                                                   InetSocketAddress resolvedPhysicalAddress) {
+        CompletableFuture<Channel> initFuture = new CompletableFuture<>();
+        ch.eventLoop().execute(() -> {
+            final ClientCnx cnx = (ClientCnx) ch.pipeline().get("handler");
+
+            if (cnx == null) {
+                initFuture
+                        .completeExceptionally(new IllegalStateException("Missing ClientCnx. This should not happen."));
+            }
```

Comment:
I wonder if this should be removed?

Instead of doing the CompletableFuture logic manually, one possibility would be to rely on Netty for that. I think that the benefit would be that all exceptions would be caught.

```
NettyFutureUtil.toCompletableFuture(ch.eventLoop().submit(() -> { .... }, ch));